inspirehep/modules/records/receivers.py

```diff
@@ -75,30 +84,74 @@ def normalize_field_categories(sender, *args, **kwargs):
 
 
-def populate_inspire_subjects(recid, json, *args, **kwargs):
-    """
-    Populate a json record before indexing it to elastic.
-    Adds a field for faceting INSPIRE subjects
+def populate_inspire_subjects(sender, json, *args, **kwargs):
+    """Populate the INSPIRE subjects before indexing.
+
+    Adds the `facet_inspire_subjects` key to the record, to be used for
+    faceting in the search interface. Valid terms on which to facet are
+    only those that come from the INSPIRE scheme.
     """
+    def _scheme_is_inspire(field):
+        return field.get('scheme') == 'INSPIRE'
+
     inspire_subjects = [
-        s['term'] for s in json.get('field_categories', [])
-        if s.get('scheme', '') == 'INSPIRE' and s.get('term')
-    ]
+        field['term'] for field in json.get('field_categories', [])
+        if _scheme_is_inspire(field) and field.get('term')]
+
     json['facet_inspire_subjects'] = inspire_subjects
 
 
-def populate_inspire_document_type(recid, json, *args, **kwargs):
-    """ Populates a json record before indexing it to elastic.
-        Adds a field for faceting INSPIRE document type
+def populate_inspire_document_type(sender, json, *args, **kwargs):
+    """Populate the INSPIRE doc type before indexing.
+
+    Adds the `facet_inspire_doc_type` key to the record, to be used for
+    faceting in the search interface. Valid document types on which to
+    facet are derived from the following algorithm:
+
+    - If the record has no `collections` key, return `[]`.
+    - Otherwise:
+      - If the record has as a primary collection a key in the following
+        table then add the corresponding value to the document types.
+        Only the first value found is added.
+
+          + --------------- + ---------------- +
+          | key             | value            |
+          + --------------- + ---------------- +
+          | published       | peer reviewed    |
+          | thesis          | thesis           |
+          | book            | book             |
+          | bookchapter     | bookchapter      |
+          | proceedings     | proceedings      |
+          | conferencepaper | conference paper |
+          | note            | note             |
+          | report          | report           |
+          | activityreport  | activity report  |
+          + --------------- + ---------------- +
+
+      - Otherwise:
+        - If the record has no `start_page` and `artid` key in any of
+          its `publication_info`s then add 'preprint' to the document types.
+      - If the record has as a primary collection a key in the following
+        table then add the corresponding value to the document types.
+        All values found are added.
+
+          + -------- + -------- +
+          | key      | value    |
+          + -------- + -------- +
+          | lectures | lectures |
+          | review   | review   |
+          + -------- + -------- +
+
```

Comment:
This is the currently implemented algorithm for assigning a doctype. It doesn't make a lot of sense. What's the real algorithm to implement here?